src/main/java/org/broadinstitute/hellbender/utils/bwa/BwaMemIndex.java

```diff
@@ -30,12 +32,55 @@
  * if that's what you want to do.
  */
 public final class BwaMemIndex implements AutoCloseable {
+
+    /**
+     * Indexing algorithms supported by Bwa.
+     */
+    public enum Algorithm {
+        RB2, IS, AUTO;
+
+        /**
+         * The string name use by the Bwa command line to denote that algorithm.
+         * @return never {@code null} and unique across algorithms.
+         */
+        public String toBwaName() {
+            return this.toString().toLowerCase();
+        }
+    }
+
     private final String indexImageFile; // stash this for error messages
     private volatile long indexAddress; // address where the index was memory-mapped (for use by C code)
     private final AtomicInteger refCount; // keep track of how many threads are actively aligning
     private final List<String> refContigNames; // the reference dictionary from the index
     private static volatile boolean nativeLibLoaded = false; // whether we've loaded the native library or not
 
+    private static final List<String> FASTA_EXTENSIONS = Collections.unmodifiableList(Arrays.asList(".fasta", ".fa"));
+
+
+    public static void indexReference(final String referenceName, final String indexPrefix, final Algorithm algo) {
+        final String referenceFileName = resolveReferenceFileName(referenceName);
+        loadNativeLibrary();
+        createReferenceIndex(referenceFileName, indexPrefix, algo.toBwaName());
+    }
+
+    private static String resolveReferenceFileName(final String referenceName) {
+        if (referenceName == null) {
+            throw new IllegalArgumentException("the reference name provided cannot be null");
+        }
+        final String referenceFileName = FASTA_EXTENSIONS.stream()
+                .filter(ext -> new File(referenceName + ext).isFile()).findFirst().orElse(referenceName);
```

Comment:
If I understand correctly, this is required for allowing `referenceName` to be a prefix. What will happen if there are `referenceName.fa` and `referenceName.fa.fasta` in the same folder? I know that it is not common, but I saw worse stuff in some folders: this will lead to unexpected behaviour for the API user if they provide `referenceName.fa` (and they know in advance that it exists)

Maybe I will force the parameter in `indexReference` to be a `File` and then perform the validation in the following lines. If you want to support the prefix behaviour, I will force the method with the `String` to be always a prefix and resolve the name there.